Source/DafnyRuntime/DafnyRuntime.py

```diff
@@ -107,22 +108,62 @@ def __lt__(self, other):
     def __le__(self, other):
         return len(self) <= len(other) and self == other[:len(self)]
 
-class Array(list):
-    @classmethod
-    def empty(cls, dims):
-        if dims <= 1:
-            return Array()
-        return Array([Array.empty(dims-1)])
+class Array:
+    class Box(list):
+        def __dafnystr__(self) -> str:
+            return '[' + ', '.join(map(string_of, self)) + ']'
+
+    def __init__(self, initValue, *dims):
+        fresh = not isinstance(initValue, Array.Box)
+        self.arr = initValue if fresh else initValue.content
+        self.dims = list(dims)
+        if fresh:
+            for i in reversed(self.dims):
+                self.arr = Array.Box([copy.deepcopy(self.arr) for _ in range(i)])
 
     def __dafnystr__(self) -> str:
-        return '[' + ', '.join(map(string_of, self)) + ']'
+        return '[' + ', '.join(map(string_of, self.arr)) + ']'
+
+    def __str__(self):
+        return self.__dafnystr__()
+
+    def length(self, i):
+        return self.dims[i] if i < len(self.dims) else None
 
     def __len__(self):
-        l = super().__len__()
-        # Hack to enable "empty" matrices
-        if l == 1 and isinstance(self[0], Array) and len(self[0]) == 0:
-            return 0
-        return l
+        return self.length(0)
+
+    def __getitem__(self, key):
+        if isinstance(key, slice):
+            return self.arr[key]
+        arr = self.arr
+        dims = self.dims
+        if not isinstance(key, Iterable):
+            key = (key,)
+        elif len(dims) < len(key):
+            return None
+        for i in key:
+            if dims[0] == 0:
+                return None
+            arr = arr[i]
+            dims = dims[1:]
+        return arr if not dims else Array(arr, dims)
```

Comment:
… as long as this line checks `len(key)` against `len(dims)`

Reply:
But, does the `else` branch ever get used by the compiler?  If not, maybe we can remove that complexity